architecture/2024-07-user-defined-types.md

```diff
@@ -0,0 +1,495 @@
+- Proxy traffic to the resource provider.
+- Interpret the result and update internal state.
+
+##### Proposal
+
+We should implement user-defined types as a resource provider with dynamic behavior. The new resource-provider could be hosted inside UCP directly or in its own microservice.
+
+### API design (if applicable)
+
+API design will be covered as part of the *Resource Type registration APIs* and *Resource Type schemas* design documents.
+
+### CLI Design (if applicable)
+
+CLI design will be covered as part of the *Tooling support for using user-defined types* design document.
+
+### Implementation Details
+
+N/A for this document. This is an overview. 
+
+### Error Handling
+
+N/A for this document. This is an overview. 
+
+## Test plan
+
+N/A for this document. This is an overview. 
+
+## Security
+
+The user-defined types feature has a few touchpoints with security.
+
+- Secret handling: As part of the API design document we will review a proposed approach for handling secrets securely. The goal of this proposal is to rely on the `secretStores` type and avoid handling secrets as part of other APIs. This will simplify the security model overall for Radius.
+
+- Strong typing for APIs and storage: As described in the [Kubernetes Structural Schema](https://kubernetes.io/blog/2019/06/20/crd-structural-schema/#towards-complete-knowledge-of-the-data-structure) blog post, an API can introduce vulnerabilities whenever it persists unvalidated data. The implementation of user-defined types will increase the level of validation and improve the overall security posture of Radius.
+
+## Compatibility (optional)
+
+Current UCP does not know the individual resource types that belong to a resource provider. It only validates the provider namespace. When we introduce the APIs for registration of resource types/providers, we should begin registering the built-in types. This can be rolled our gradually without breaking users.
+
+## Monitoring and Logging
+
+The Dynamic-RP design document will cover monitoring and logging. We don't anticipate any new kinds of telemetry, but if we build a new component we must add the boilerplate to produce the telemetry we already support.
+
+## Development plan
+
+We can implement user-defined types through a staged development plan. This plan is designed to enable the user-defined type end-to-end early to unblock functional testing and community feedback. After user-defined types reach MVP status, we shift focus to enhancing them and getting the community involved.
+
+Stages are described in linear order, but many of them can be parallelized. I chose what I felt was a sensible progression, but it's not strictly defined.
+
+### Phase 1: Enabling the UDT end-to-end
+
+#### Stage 1: Type Registration APIs
+
+- Define API for CRUDL operations and resource providers and resource types.
+- Add `rad` CLI commands `rad resourceprovider xyz` and `rad resourcetype xyz`.
```

Comment:
I think, we also talked about having a resource type being a parent of another resource type. I know that needs further design. We can also keep that in mind when we are working on this part of the work.

Reply:
This will be covered in another design doc. This is the overall phasing of work.

I'd actually really liked to declare this out of scope for now until there's demand from users.